promql/engine.go

```diff
@@ -396,20 +398,21 @@ func (ng *Engine) validateOpts(expr parser.Expr) error {
 	parser.Inspect(expr, func(node parser.Node, path []parser.Node) error {
 		switch n := node.(type) {
 		case *parser.VectorSelector:
-			if n.Timestamp != nil {
-				validationErr = errors.New("@ modifier is disabled")
+			if n.Timestamp != nil || n.Preprocessor == parser.START || n.Preprocessor == parser.END {
+				validationErr = ErrValidationAtModifierDisabled
 				return validationErr
 			}
```

Comment:
can you return directly the error?

Reply:
Indeed. I think you don't need the `validationErr` variable at all (L397). Just return the error whenever it occurs, and return `nil` in the end. No?

Reply:
Yeah, some cleanup leftovers :)

Reply:
Oh wait, no. The parser.Inspect does not return any error, the returned error here just tells not to walk the AST further (notice that its inside parser.Inspect). So I am having this variable to use directly instead of a boolean.

Reply:
:+1:

Reply:
As a lover of closures, I should have seen that…